File: proj_lab1/oolab/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;

import java.util.Objects;

public class Animal implements WorldElement {
    private MapDirection direction = MapDirection.NORTH;
    private Vector2d position;

    public Animal(){
        this(new Vector2d(2,2));
    }

    public Animal(Vector2d position){
        this.position = position;
    }

    public MapDirection getDirection() {
        return direction;
    }

    public Vector2d getPosition() {
        return position;
    }

    //na potrzeby komparatora
    public int getPositionX(){
        return this.position.getX();
    }

    //na potrzeby komparatora
    public int getPositionY(){
        return this.position.getY();
    }

    //zmienione na potrzeby etykiety
    @Override
    public String toString() {
        return "Z " + position.toString();
    }

    public boolean isAt(Vector2d position){
        return this.position.equals(position);
    }

    public void move(MoveValidator validator, MoveDirection direction){
        Vector2d newPosition = position;
        switch(direction){
            case LEFT -> this.direction = this.direction.previous();
            case RIGHT -> this.direction = this.direction.next();
            case FORWARD, BACKWARD -> newPosition = direction == MoveDirection.FORWARD ? position.add(this.direction.toUnitVector()) : position.subtract(this.direction.toUnitVector());
        }
        if (validator.canMoveTo(newPosition)) {
            this.position = newPosition;
        }
    }

    @Override
    public String imagePath(){
        return switch(this.direction){
            case NORTH -> "images/up.png";
            case EAST -> "images/right.png";
            case SOUTH -> "images/down.png";
            case WEST -> "images/left.png";
        };
    }

    //na potrzeby getElements()
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Animal animal = (Animal) o;
        return Objects.equals(position, animal.position);
    }

    //na potrzeby getElements()
    @Override
    public int hashCode() {
        return Objects.hash(position);
    }
}

File: proj_lab2/oolab/src/main/kotlin/agh/ics/oop/model/Animal.kt
package agh.ics.oop.model

data class Animal (var position: Vector2d = Vector2d(2, 2)) {
    private var direction = MapDirection.NORTH

    fun getPositionX(): Int = position.getX()
    fun getPositionY(): Int = position.getY()

    override fun toString(): String = "Z $position"

    fun isAt(position: Vector2d): Boolean = this.position == position

    fun move(validator: IMoveValidator, direction: MoveDirection) {
        var newPosition = position
        val unitVectorMovement = this.direction.toUnitVector()
        when (direction) {
            MoveDirection.LEFT -> this.direction = this.direction.previous()
            MoveDirection.RIGHT -> this.direction = this.direction.next()
            MoveDirection.FORWARD -> newPosition += unitVectorMovement
            MoveDirection.BACKWARD -> newPosition -= unitVectorMovement
        }
        if (validator.canMoveTo(newPosition)) {
            position = newPosition
        }
    }
}


